File: IT/Oggetti/Esercizi/Python/esercizio29.py
# file esercizio29.py

from __future__ import print_function
from math import pi

# la classe Circle rappresenta una circonferenza

class Circle(object):
	# ogni circonferenza viene definita dal suo raggio
	def __init__(self, r):
		self.r = r

	# area del cerchio = Pi greco * raggio al quadrato
	def area(self):
		return pi * self.__r * self.__r

	def getR(self):
		return self.__r

	def setR(self, r):
		self.__r = r

	r = property(getR, setR)

# la classe Sphere rappresenta una sfera

class Sphere(object):
	# la sfera viene definita dal suo cerchio massimo
	def __init__(self, circle):
		self.__circle = circle

	# volume della sfera: (4 * 3.14 * r^3) / 3
	def volume(self):
		return (4 * self.__circle.area() * self.__circle.r) / 3

circle = Circle(r=2.0)

print("Area circonferenza alla base: {:.2f} centimetri quadrati.".format(circle.area()))

sphere = Sphere(circle=circle)

print("Volume sfera: {:.2f} centimetri cubi.".format(sphere.volume())) 

File: IT/Oggetti/Esercizi/C++/esercizio29.cpp
/* file esercizio29.cpp */

#include <iostream>
#include <iomanip>
#include <cmath>

using std::cout;
using std::endl;
using std::string;
using std::setprecision;
using std::ostream;

/* la classe Circle rappresenta una circonferenza */

class Circle {
	/* ogni circonferenza viene definita dal suo raggio */
	private:
		double r;

	public:
		Circle(double);
		double area();
		double getR();
		void setR(double);
};

Circle::Circle(double r) {
	this->setR(r);
}

/* area del cerchio = Pi greco * raggio al quadrato */
double Circle::area() {
	return M_PI * this->r * this->r;
}

double Circle::getR() {
	return this->r;
}

void Circle::setR(double r) {
	this->r = r;
}

/* la classe Sphere rappresenta una sfera */

class Sphere {
	/* la sfera viene definita dal suo cerchio massimo */
	private:
		Circle *circle;

	public:
		Sphere(Circle*);
		double volume();
};

Sphere::Sphere(Circle *circle) {
	this->circle = circle;
}

/* volume della sfera: (4 * 3.14 * r^3) / 3 */
double Sphere::volume() {
	return (4 * this->circle->area() * this->circle->getR()) / 3;
}

int main()
{
	Circle *circle = new Circle(2.0);

	cout << "Area circonferenza massima: " << setprecision(4) << circle->area() << " centimetri quadrati." << endl;

	Sphere *solid = new Sphere(circle);

	cout << "Volume sfera: " << setprecision(4) << solid->volume() << " centimetri cubi." << endl;

	delete circle;
	delete solid;
}

